src/pkg/golang/build.go

```diff
@@ -204,19 +232,107 @@ func (c Environ) GoCmd(gocmd string, args ...string) *exec.Cmd {
 	return cmd
 }
 
-// Version returns the Go version string that runtime.Version would return for
-// the Go compiler in this environ.
-func (c Environ) Version() (string, error) {
+// resolve absolute path of go-compiler option from PATH
+func (c *Environ) resolveCompiler() error {
+	if c.Compiler.Path != "" {
+		fname, err := exec.LookPath(string(c.Compiler.Path))
+		if err == nil {
+			fname, err = filepath.Abs(fname)
+		}
+		if err != nil {
+			return fmt.Errorf("build: %v", err)
+		}
+		c.Compiler.Path = fname
+	}
+	return nil
+}
+
+// runs GoCmd("version") and parse/caches output, to environ.Compiler
+func (c *Environ) InitCompiler() error {
+
+	if c.Compiler.IsInit {
+		return nil
+	}
+
+	c.resolveCompiler()
+
 	cmd := c.GoCmd("version")
 	v, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", err
+		return err
 	}
+
+	efmt := "go-compiler 'version' output unrecognized: %v"
 	s := strings.Fields(string(v))
-	if len(s) < 3 {
-		return "", fmt.Errorf("unknown go version, tool returned weird output for 'go version': %v", string(v))
+	if len(s) < 1 {
+		return fmt.Errorf(efmt, string(v))
 	}
-	return s[2], nil
+
+	compiler := c.Compiler
+	compiler.Identifier = s[0]
+	compiler.VersionOutput = string(v)
+	compiler.IsInit = true
+
+	switch compiler.Identifier {
+
+	case "go":
+		if len(s) < 3 {
+			return fmt.Errorf(efmt, string(v))
+		}
+		compiler.Version = s[2]
+		compiler.VersionGo = s[2]
+
+	case "tinygo":
+		// e.g. "tinygo version 0.33.0 darwin/arm64 (using go version go1.22.2 and LLVM version 18.1.2)"
+		if len(s) < 8 {
+			return fmt.Errorf(efmt, string(v))
+		}
+		compiler.Version = s[2]
+		compiler.VersionGo = s[7]
+
+		// Fetch additional go-build-tags from tinygo
+		// package fetch needs correct tags to prune
+		cmd := c.GoCmd("info", "-json")
+		infov, err := cmd.CombinedOutput()
+		if err != nil {
+			return err
+		}
+		var info map[string]interface{}
+		err = json.Unmarshal(infov, &info)
+		if err != nil {
+			return err
+		}
+		tags := []string{}
+		unique := make(map[string]bool)
+		addUnique := func(tag string) {
+			if unique[tag] {
+				return
```

Comment:
tags := make(map[string]struct{})
for _, tag := range c.BuildTags {
  tags[tag] = struct{}{}
}
for _, tag := range info["build_tags"].([]string) {
  tags[tag] = struct{}{}
}
c.BuildTags = maps.Keys(tags) // import "maps"